calyx-py/test/correctness/queues/binheap.py

```diff
@@ -29,107 +31,369 @@ def insert_swap(prog, name, width, len, idx_w):
+
+    current_idx = comp.reg(factor)
+    current_rank = comp.reg(32)
+
+    parent_idx = comp.reg(factor)
+    parent_rank = comp.reg(32)
 
-    sub = comp.sub(4)
-    rsh = comp.rsh(4)
+    child_l_idx = comp.reg(factor)
+    child_l_rank = comp.reg(32)
 
-    size = comp.reg(4)  # Current size
-    parent_idx = comp.reg(4)
-    parent_val = comp.reg(64)
-    child_idx = comp.reg(4)
-    child_val = comp.reg(64)
+    child_r_idx = comp.reg(factor)
+    child_r_rank = comp.reg(32)
 
-    read_parent = comp.mem_load_d1(mem, parent_idx.out, parent_val, "read_parent")
-    read_child = comp.mem_load_d1(mem, child_idx.out, child_val, "read_child")
+    # current_idx := 0
+    set_idx_zero = comp.reg_store(current_idx, 0, "set_idx_zero")
 
+    # current_idx := size
+    set_idx_size = comp.reg_store(current_idx, size.out, "set_idx_size")
+    
+    # current_idx := child_l_idx
+    set_idx_child_l = comp.reg_store(current_idx, child_l_idx.out, "set_idx_child_l")
+
+    # current_idx := child_r_idx
+    set_idx_child_r = comp.reg_store(current_idx, child_r_idx.out, "set_idx_child_r")
+    
+    # current_idx := parent_idx
+    set_idx_parent = comp.reg_store(current_idx, parent_idx.out, "set_idx_parent")
+
+    # err := 1
+    raise_err = comp.reg_store(err, 1, "raise_err") 
+    
+    sub = comp.sub(factor)
+    rsh = comp.rsh(factor)
     with comp.group("find_parent_idx") as find_parent_idx:
-        # Find the parent of the `child_idx`th element and store it in `parent_idx`.
-        # parent_idx := floor((child_idx − 1) / 2)
-        sub.left = child_idx.out
+        # Find the parent of the `current_idx`th element and store it in `parent_idx`.
+        # parent_idx := floor((current_idx − 1) / 2)
+        sub.left = current_idx.out
         sub.right = 1
         rsh.left = sub.out
-        rsh.right = cb.const(4, 1)
+        rsh.right = cb.const(factor, 1)
         parent_idx.in_ = rsh.out
         parent_idx.write_en = cb.HI
         find_parent_idx.done = parent_idx.done
 
-    set_child_idx = comp.reg_store(child_idx, size.out)  # child_idx := size
-    store_new_val = comp.mem_store_d1(
-        mem, child_idx.out, value, "store_new_val"
-    )  # mem[child_idx] := value
-    incr_size = comp.incr(size)
-    child_lt_parent = comp.lt_use(child_val.out, parent_val.out)
-    bubble_child_idx = comp.reg_store(child_idx, parent_idx.out, "bubble_child_idx")
+    add_1 = comp.add(factor)
+    add_2 = comp.add(factor)
+    lsh = comp.lsh(factor)
+    with comp.group("find_child_idx") as find_child_idx:
+        # Find the children of `current_idx`th element and store it in child_l_idx and child_r_idx.
+        # child_l_idx := (2 * current_idx) + 1
+        # child_r_idx := (2 * current_idx) + 2
+        lsh.left = current_idx.out
+        lsh.right = cb.const(factor, 1)
+        add_1.left = 1
+        add_1.right = lsh.out
+        child_l_idx.write_en = cb.HI
+        child_l_idx.in_ = add_1.out
+        add_2.left = child_l_idx.done @ 1
```

Comment:
Minor, but this could be restructured into two groups that can safely run in parallel.

Reply:
That's true! I originally wrote it this way because then I could have one less left shifter (i.e. I wouldn't need two shifters performing the computation for `2 * current_idx` twice). 

Since everything going on here is combinational, do you think it matters whether I factor it out into two groups running in parallel?